Fix using --help, --verbose, etc.

ehuss opened the pull request:
When running `cargo fmt -- --help`, no output would be displayed.

Closes rust-lang/cargo#7027.

ehuss commented:
This solution is a little sloppy, but I think anything else would require a much larger change.

scampi commented on src/cargo-fmt/main.rs:
```diff
@@ -345,6 +346,11 @@ fn run_rustfmt(
             h
         });
 
+    let s = "2018".to_string();
```
I think it'd be better to use the default of `Edition`.

ehuss replied:
I'm not sure I understand.  `Edition` doesn't implement anything like Display to get a string.  Also, this bin currently doesn't import anything from rustfmt, so that would be a distinct change.

scampi replied:
Indeed. However I don't really like scattering defaults all over, especially here where it'd be to get around a piece of code. With a bit of refactoring, it'll be nicer:

- we are already executing `rustfmt` without the need for edition in order to get the version:
     https://github.com/rust-lang/rustfmt/blob/466ddc9aba88dd33e28226642223c5285c47336a/src/cargo-fmt/main.rs#L87-L89 
- we could refactor a bit `get_version` so that it can execute other option beside `version`
    https://github.com/rust-lang/rustfmt/blob/466ddc9aba88dd33e28226642223c5285c47336a/src/cargo-fmt/main.rs#L120 
- this way the `if` block wouldn't be needed since any of the options (`--help`, `-V` ...) would be handled by the refactored `get_version` method https://github.com/rust-lang/rustfmt/blob/466ddc9aba88dd33e28226642223c5285c47336a/src/cargo-fmt/main.rs#L145-L149

How does that sound ?

ehuss commented:
> Can you add a test ?

Sure.  However, there aren't any tests for `cargo fmt` that I can find.  I could set up an integration test to launch it.  It won't be completely trivial (it might require things like PATH modification), though.  Is that what you're thinking of?

scampi commented:
@ehuss re test: I never did so myself, but is it possible to create a method which returns the binary's name, conditionally compiling the method on whether or not a test is run ?

Instead of below, do `Command::new(get_binary())`:

https://github.com/rust-lang/rustfmt/blob/466ddc9aba88dd33e28226642223c5285c47336a/src/cargo-fmt/main.rs#L121

```rust
#[cfg(test)]
fn get_binary() -> String {
    // absolute path should be returned
    "./target/debug/rustfmt"
}

#[cfg(not(test))]
fn get_binary() -> String {
    "rustfmt"
}
```

This would mean that `cargo build` needs to have run before, but anyway this is already required by some tests:

https://github.com/rust-lang/rustfmt/blob/dbac28b4b4f2adb135104593365943b3eab24685/src/test/mod.rs#L1074-L1078

ehuss commented:
I have updated with the different approach, and added some tests.  I just used the standard current_exe trick for now.  I'm actually working on adding a first-class way for tests to find the binaries in cargo, so hopefully sometime soon that won't be necessary anymore.

scampi commented on tests/cargo-fmt/main.rs:
```diff
@@ -0,0 +1,70 @@
+use std::env;
+use std::process::Command;
+
+/// Run the cargo-fmt executable and return its output.
```
Can you add comment that the tuple corresponds to `(stdout, stderr)` of the execution ?

topecongiro commented:
Thanks!
